datalad/interface/publish_collection.py

```diff
@@ -0,0 +1,312 @@
+                raise e
+
+            handle_publisher(None, handle=handle_loc,
+                             url=baseurl + '/' + handle_name)
+
+        # TODO: check success => go on with collection
+
+        # prepare publish branch in local collection:
+        # check for existing publish branches:
+        from random import choice
+        from string import letters
+        p_branch = "publish_" + ''.join(choice(letters) for i in xrange(6))
+        local_collection_repo.git_checkout(p_branch, '-b')
+
+        importer = CustomImporter('Collection', 'Collection', DLNS.this)
+        importer.import_data(local_collection_repo.path)
+        graphs = importer.get_graphs()
+        orig_uri = graphs[REPO_STD_META_FILE[0:-4]].value(predicate=RDF.type,
+                                                          object=DLNS.Handle)
+        # correct collection uri
+        new_uri = URIRef(collection_url)
+        for graph_name in graphs:
+            for p, o in graphs[graph_name].predicate_objects(subject=orig_uri):
+                graphs[graph_name].remove((orig_uri, p, o))
+                graphs[graph_name].add((new_uri, p, o))
+
+        # correct handle uris in hasPart statements:
+        replacements = []
+        from datalad.support.collection import Collection
+        from datalad.support.collectionrepo import CollectionRepoBackend
+        col_meta = Collection(CollectionRepoBackend(local_collection_repo))
+        for o in graphs[REPO_STD_META_FILE[0:-4]].objects(subject=new_uri,
+                                                          predicate=DCTERMS.hasPart):
+            from os.path import basename
+            path = urlparse(o).path
+            if exists(path):
+                # local handle
+                # retrieve name for that uri:
+                # Note: That's an experimental implementation
+                hdl_name = None
+                for key in col_meta:
+                    if urlparse(col_meta[key].url).path == path:
+                        hdl_name = col_meta[key].name
+                if hdl_name is None:
+                    raise RuntimeError("No handle found for path '%s'." % path)
+
+                o_new = URIRef(baseurl + '/' + hdl_name)
+                replacements.append((o, o_new))
+            else:
+                # TODO: what to do? We have a locally not available handle
+                # in that collection, that therefore can't be published.
+                # Just skip for now and assume uri simply doesn't change.
+                continue
+        for o, o_new in replacements:
+            graphs[REPO_STD_META_FILE[0:-4]].remove((new_uri, DCTERMS.hasPart, o))
+            graphs[REPO_STD_META_FILE[0:-4]].add((new_uri, DCTERMS.hasPart, o_new))
+
+        # TODO: add commit reference
+
+        importer.store_data(local_collection_repo.path)
+        [local_collection_repo.git_add(graph_name + '.ttl')
+         for graph_name in graphs]
```

Comment:
minor whining: abuse of list comprehensions -- why bother creating a list (instead of straight for loop) if output is not to be consumed?